File: leetcode/golang/LCR_146_shun_shi_zhen_da_yin_ju_zhen_lcof.go
func spiralArray(array [][]int) []int {
    if len(array) == 0 || len(array[0]) == 0 {
        return []int{}
    }
    l, r, t, b := 0, len(array[0]) - 1, 0, len(array) - 1
    var res []int
    for true {
        for i := l; i <= r; i++ {
            res = append(res, array[t][i])
        }
        t++
        if t > b {
            break
        }
        for i := t; i <= b; i++ {
            res = append(res, array[i][r])
        } 
        r--
        if r < l {
            break
        }
        for i := r; i >= l; i-- {
            res = append(res, array[b][i])
        }
        b--
        if b < t {
            break
        }
        for i := b; i >= t; i-- {
            res = append(res, array[i][l])
        }
        l++
        if l > r {
            break
        }
    }
    return res
}

File: leetcode/c++/LCR_146_shun_shi_zhen_da_yin_ju_zhen_lcof.cpp
// https://leetcode.cn/problems/shun-shi-zhen-da-yin-ju-zhen-lcof/description

#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    vector<int> spiralArray(vector<vector<int>>& array) {
        if (array.empty())
            return {};
        int top = 0, bottom = array.size() - 1, left = 0, right = array[0].size() - 1;
        vector<int> res;
        while (true) {
            for (int i = left; i <= right; i++) res.push_back(array[top][i]);
            if (++top > bottom) break;
            for (int i = top; i <= bottom; i++) res.push_back(array[i][right]);
            if (--right < left) break;
            for (int i = right; i >= left; i--) res.push_back(array[bottom][i]);
            if (--bottom < top) break;
            for (int i = bottom; i >= top; i--) res.push_back(array[i][left]);
            if (++left > right) break;
        }
        return res;
    }
};

// 模拟，O(m * n)，O(1)